### ISSUE TITLE
Add API for switching between GPU and CPU

### ISSUE THREAD

**Issue by matthewfeickert** [scikit-hep/pyhf#896]
# Description

Given the desire to be able to test the performance of the CPU and GPU backends, and that by default the TensorFlow, PyTorch, and JAX backends all attempt to run in GPU mode, it would be helpful to add the ability to set the "mode" (if you will) to both the Python and CLI API.

@coolalexzb can probably also give relevant advice on what is needed in this area.

**Comment by matthewfeickert** [comment]
With regards to JAX, it seems that the [recommended way to do this is by setting the `JAX_PLATFORM_NAME` environment variable](https://github.com/google/jax/issues/696#issuecomment-642827038). So an example of this might be

```python
import os
mode = user_input_mode.lower() # "gpu" or "cpu"
os.environ["JAX_PLATFORM_NAME"] = mode
```

but for TensorFlow it might be a bit more complicated as to run in CPU mode I've seen people suggest simply masking the existence of GPUs from the visible devices

```python
import tensorflow as tf
tf.config.set_visible_devices([], "GPU")
```

but this isn't great as if a user has switch from GPU to CPU, we want the user to be able to go back from CPU to GPU and that requires being able to find the GPUs in

```python
tf.config.get_visible_devices()
# example output
# [PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU'),
# PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]
```

**Comment by coolalexzb** [comment]
To [test the performance of the CPU and GPU backends](https://github.com/pyhf/pyhf-benchmark), we will firstly use `pyhf.set_backend(backend)` in [run.py](https://github.com/pyhf/pyhf-benchmark/blob/master/scripts/run.py) to set computation backend. From my suggestion, we can add CPU or GPU option as a parameter in `pyhf.set_backend` function and add "set mode" code in `_OptimizerRetriever` class.

**Comment by coolalexzb** [comment]
As you mentioned 

> we want the user to be able to go back from CPU to GPU

Can you show me how users will interact to switch between CPU and GPU @matthewfeickert

**Comment by coolalexzb** [comment]
For PyTorch, from [Easy way to switch between CPU and cuda](https://github.com/pytorch/pytorch/issues/1668)

```
mode = user_input_mode.lower()
if mode == "gpu":
    import torch.cuda as t
else:
    import torch as t

x = t.zeros()
y = Variable(x)
```

or from [Porting PyTorch code from CPU to GPU](https://stackoverflow.com/questions/46704352/porting-pytorch-code-from-cpu-to-gpu)

```
mode = user_input_mode.lower()
net = YouNetworkClass()
device = torch.device("cuda:0" if mode == "gpu" else "cpu")
net.to(device)
```

**Comment by coolalexzb** [comment]
As for Tensorflow,  maybe we can use the following [approach](https://stackoverflow.com/questions/37660312/how-to-run-tensorflow-on-cpu), which is similar to the approach mentioned to JAX before. 

```
import os
mode = user_input_mode.lower()
if mode == "cpu":
  os.environ["CUDA_VISIBLE_DEVICES"] = "-1"  
else:
   os.environ["CUDA_VISIBLE_DEVICES"] = ""  
import tensorflow as tf
```

**Comment by lukasheinrich** [comment]
hi @coolalexzb 

I think we might be able to avoid going through env vars

e.g. in torch you can put tensors on GPU devices through the `.cuda()` method

<img width="292" alt="screenshot" src="https://user-images.githubusercontent.com/2318083/84833473-a23bc900-b02f-11ea-9b54-3e1a2a6d46a0.png">


so in the `astensor` method  of the pytorch backend you could just append a `.cuda()` call

**Comment by lukasheinrich** [comment]
for TF this would work

<img width="465" alt="screenshot" src="https://user-images.githubusercontent.com/2318083/84834051-d82d7d00-b030-11ea-9d2d-77ff2ca1b55c.png">

**Comment by matthewfeickert** [comment]
> Hi, is there any progress or status on when to expect this functionality?

@nhartman94 This isn't on the roadmap for `v0.6.2` but if this is of high interest we could look at putting it on for `v0.7.0`. Do you currently have a use case where you're running normally on GPUs and then want to switch to running on CPUs? This feature was more of interest to the dev team for us to perform benchmarking studies. Do you have a different use case here?

**Comment by nhartman94** [comment]
Hi @matthewfeickert thanks, for answering! I might have more trivial use case, I was just wondering how to minimize the likelihood using the GPU w/ the deep learning backends. I don't think I need to switch and forth. How would I go about just "running normally on GPUs"?

**Comment by kratsg** [comment]
> How would I go about just "running normally on GPUs"?

just enable GPU for the backend you'd want to use, such as pytorch or tensorflow or jax -- and you would need to refer to the corresponding documentation on how to check that you're set up to run with a GPU. `pyhf` does not do anything special to do this (right now).

**Comment by nhartman94** [comment]
Ok, thanks, do I need to transfer the input histograms onto the gpu?

**Comment by kratsg** [comment]
> Ok, thanks, do I need to transfer the input histograms onto the gpu?

This should be handled for you by the corresponding backend when the tensors are created from the workspace.

**Comment by matthewfeickert** [comment]
> (For context, I was thinking about where the step of transferring the tensors to the gpu when training pytorch NNs happens analogously when fitting the likelihood.)
> ...
> This should be handled for you by the corresponding backend when the tensors are created from the workspace.

For additional context, at the moment `pyhf` just places everything on the first GPU device that is available to it for the specified backend. We currently don't support the ability to specify how to split data out across multiple GPUs.

**Comment by nhartman94** [comment]
Ok, thank you both for the insight. I tried this logging into a machine and minimizing the likelihood, but while it's running, and I run nvidia-smi in a terminal, it doesn't seem like the gpu is being use.  Is there a way to tell if it's using the gpu, or this is just a lightweight enough application that it's not showing up?

**Comment by lukasheinrich** [comment]
Can you check `.device()` for tthe following (assuming jax)

```
>>> import pyhf
>>> pyhf.set_backend('jax')
>>> a = pyhf.tensorlib.astensor([1.,2.,3.])
>>> a.device()
CpuDevice(id=0)
```

**Comment by nhartman94** [comment]
<img width="225" alt="Screen Shot 2021-04-23 at 12 12 13 AM" src="https://user-images.githubusercontent.com/14224584/115791373-98f65800-a3c8-11eb-8d66-2c892c41703b.png">

Ok, thanks. I'm on a machine w/ a gpu, but the tensor is getting created on the cpu (I think is what this is saying?)

**Comment by matthewfeickert** [comment]
> Ok, thank you both for the insight. I tried this logging into a machine and minimizing the likelihood, but while it's running, and I run nvidia-smi in a terminal, it doesn't seem like the gpu is being use. Is there a way to tell if it's using the gpu, or this is just a lightweight enough application that it's not showing up?

@nhartman94 in addition to @lukasheinrich's checks you can also use my homebrewed GPU detection scripts: https://github.com/matthewfeickert/nvidia-gpu-ml-library-test

**Comment by matthewfeickert** [comment]
> Ok, thanks. I'm on a machine w/ a gpu, but the tensor is getting created on the cpu (I think is what this is saying?)

@nhartman94 yeah. Seems that we still don't have Issue #1244 taken care of.

**Comment by nhartman94** [comment]
Oh, ok thanks haha

**Comment by matthewfeickert** [comment]
> Oh, ok thanks haha - this makes sense now :)

@nhartman94 Note that Issue #1244 only affects PyTorch, so you should be good to go for JAX.

Example:

```
(pyhf-dev) $ pip list | grep jax
jax                           0.2.12
jaxlib                        0.1.65+cuda101
(pyhf-dev) $ curl -sL https://raw.githubusercontent.com/matthewfeickert/nvidia-gpu-ml-library-test/main/jax_detect_GPU.py | python
XLA backend type: gpu

Number of GPUs found on system: 1

Active GPU index: 0
Active GPU name: GeForce GTX 1650 with Max-Q Design
(pyhf-dev) $ python
Python 3.8.6 (default, Jan  5 2021, 00:14:15) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyhf
>>> pyhf.set_backend("jax")
>>> a = pyhf.tensorlib.astensor([1.,2.,3.])
>>> a.device()
GpuDevice(id=0, process_index=0)
```

Sorry this is still an Issue, but I'll ping you over on Issue #1244 for some timeline discussion.

**Comment by nhartman94** [comment]
Ok - thank you so much for all the help :)

For some reason, I still seemed to be having issues w/ not picking them up with the TensorFlow backend - but when I did jax ( and following your instructive [link](https://github.com/matthewfeickert/nvidia-gpu-ml-library-test) @matthewfeickert for making sure I had jax w/ the gpu backend installed, I can see the gpu process running while the likelihood is minimizing 😄
